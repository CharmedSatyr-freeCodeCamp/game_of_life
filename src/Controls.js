import React, { Component } from 'react';
import PropTypes from 'prop-types';
import * as c from './constants/constants';

let duration;
export default class Controls extends Component {
  constructor(props) {
    super(props);
    this.clear = this.clear.bind(this);
    this.makeGrid = this.makeGrid.bind(this);
    this.start = this.start.bind(this);
    this.stop = this.stop.bind(this);
  }
  clear() {
    this.props.clear();
    clearInterval(duration);
  }
  makeGrid() {
    this.props.makeGrid();
  }
  start() {
    duration = setInterval(() => {
      this.props.nextGen();
    }, c.duration);
  }
  stop() {
    clearInterval(duration);
  }
  render() {
    return (
      <div>
        <button onClick={this.start}>Start</button>
        <button onClick={this.stop}>Stop</button>
        <button onClick={this.clear}>Clear</button>
        <button onClick={this.makeGrid}>Randomize</button>
      </div>
    );
  }
}

Controls.propTypes = {
  clear: PropTypes.func.isRequired,
  makeGrid: PropTypes.func.isRequired,
  nextGen: PropTypes.func.isRequired,
};
