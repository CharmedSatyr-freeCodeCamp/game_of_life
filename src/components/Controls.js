import React, { Component } from 'react';
import PropTypes from 'prop-types';
import * as c from '../constants/constants';

// Font Awesome
import { library } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faPause,
  faPlay,
  faRandom,
  faStepForward,
  faTimes,
} from '@fortawesome/free-solid-svg-icons';
library.add(faPause, faPlay, faRandom, faStepForward, faTimes);

export default class Controls extends Component {
  constructor(props) {
    super(props);
    this.clear = this.clear.bind(this);
    this.pause = this.pause.bind(this);
    this.play = this.play.bind(this);
    this.random = this.random.bind(this);
    this.step = this.step.bind(this);

    this.state = { selected: null };
  }
  clear() {
    if (this.state.selected !== 'clear') {
      clearInterval(this.duration);
      this.props.clear();
      this.setState({ selected: 'clear' });
    }
  }
  pause() {
    if (
      this.state.selected !== 'pause' &&
      this.state.selected !== 'clear' &&
      this.state.selected !== 'random'
    ) {
      clearInterval(this.duration);
      this.setState({ selected: 'pause' });
    }
  }
  play() {
    if (this.state.selected !== 'clear' && this.state.selected !== 'play') {
      clearInterval(this.duration);
      this.duration = setInterval(() => {
        this.props.nextGen();
      }, c.duration);
      this.setState({ selected: 'play' });
    }
  }
  random() {
    clearInterval(this.duration);
    this.props.makeGrid();
    this.setState({ selected: 'random' });
  }
  step() {
    if (this.state.selected !== 'clear' && this.state.selected !== 'play') {
      this.props.nextGen();
    }
  }
  componentDidMount() {
    this.play();
  }
  render() {
    const { selected } = this.state;
    return (
      <div className="controls">
        {/* STEP */}
        <button
          className={selected !== 'random' && selected !== 'pause' && 'disabled'}
          onClick={this.step}
        >
          <FontAwesomeIcon icon="step-forward" />
          &nbsp;Step
        </button>

        {/* PLAY */}
        <button
          className={`${selected === 'play' && 'selected disabled'} ${selected === 'clear' &&
            'disabled'}`}
          onClick={this.play}
        >
          <FontAwesomeIcon icon="play" />
          &nbsp;Play
        </button>

        {/* PAUSE */}
        <button
          className={`${selected === 'pause' && 'selected disabled'} ${selected !== 'play' &&
            'disabled'}`}
          onClick={this.pause}
        >
          <FontAwesomeIcon icon="pause" />
          &nbsp;Pause
        </button>

        {/* CLEAR */}
        <button className={selected === 'clear' && 'selected disabled'} onClick={this.clear}>
          <FontAwesomeIcon icon="times" />
          &nbsp;Clear
        </button>

        {/* RANDOM */}
        <button className={selected === 'random' && 'selected'} onClick={this.random}>
          <FontAwesomeIcon icon="random" />
          &nbsp;Random
        </button>
      </div>
    );
  }
}

Controls.propTypes = {
  clear: PropTypes.func.isRequired,
  makeGrid: PropTypes.func.isRequired,
  nextGen: PropTypes.func.isRequired,
};
