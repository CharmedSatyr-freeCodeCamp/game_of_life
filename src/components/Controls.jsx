import React from 'react';

const Controls = ({onClickStart, onClickPause}) => (
    <div>
        <button onClick={onClickStart}>Start</button>
        <button onClick={onClickPause}>Pause</button>
        <button>Random</button>
    </div>
);

export default Controls;
