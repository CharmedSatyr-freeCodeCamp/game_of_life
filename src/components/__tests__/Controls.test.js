import React from 'react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

import Controls from '../Controls';

configure({ adapter: new Adapter() });

const raf = jest.spyOn(window, 'requestAnimationFrame');
const caf = jest.spyOn(window, 'cancelAnimationFrame');

describe('<Controls /> component', () => {
  const props = { clear: jest.fn(), makeGrid: jest.fn(), nextGen: jest.fn() };
  const controls = mount(<Controls {...props} />);
  const clear = controls.find('.clear');
  const pause = controls.find('.pause');
  const play = controls.find('.play');
  const random = controls.find('.random');
  const step = controls.find('.step');

  it('renders without crashing', () => {
    expect(controls.exists()).toBeTruthy();
  });

  it('renders correctly', () => {
    const tree = renderer.create(<Controls {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('`Clear` button', () => {
    it('runs `clear` when state is not `clear`', () => {
      clear.simulate('click');
      expect(props.clear).toHaveBeenCalledTimes(1);
    });

    it('does nothing if state is `clear`', () => {
      clear.simulate('click');
      expect(props.clear).not.toHaveBeenCalledTimes(2);
    });

    it('cancels the animation frame for the component `requestID`', () => {
      expect(caf).toHaveBeenCalledTimes(1);
    });

    it('sets state to `clear`', () => {
      expect(controls.state('selected')).toBe('clear');
    });
  });

  describe('`Pause` button', () => {
    it('does nothing if state is `clear`, `pause`, or `random`', () => {
      expect(controls.state('selected')).toBe('clear');
      pause.simulate('click');
      expect(controls.state('selected')).toBe('clear');
    });

    it('calls `cancelAnimationFrame` on the component `requestID`', () => {
      play.simulate('click');
      pause.simulate('click');
      expect(caf).toHaveBeenCalledTimes(2);
    });

    it('sets state to `pause`', () => {
      expect(controls.state('selected')).toBe('pause');
    });
  });

  describe('`Play` button', () => {
    it('runs `nextGen` when clicked', () => {
      play.simulate('click');
      // Previously called in `renderer`, `componentDidMount` and `Pause` test
      expect(props.nextGen).toHaveBeenCalledTimes(4);
    });

    it('assigns a recursive `requestAnimationFrame` to the component as a `requestID`', () => {
      // Previously called in `renderer`, `componentDidMount` and `Pause` test
      expect(raf).toHaveBeenCalledTimes(4);
    });

    it('sets state to `play`', () => {
      expect(controls.state('selected')).toBe('play');
    });
  });

  describe('`Random` button', () => {
    it('calls `cancelAnimationFrame` on the component `requestID`', () => {
      random.simulate('click');
      expect(caf).toHaveBeenCalledTimes(3);
    });

    it('assigns a recursive `requestAnimationFrame` to the component as a `requestID`', () => {
      expect(props.makeGrid).toHaveBeenCalledTimes(1);
    });

    it('sets state to `random`', () => {
      expect(controls.state('selected')).toBe('random');
    });
  });

  describe('`Step` button', () => {
    it('does nothing if the state is `play`', () => {
      // Calls nextGen again (4x above -> 5x)
      play.simulate('click');
      // Does not call nextGen again (expect still 5x)
      step.simulate('click');
      expect(props.nextGen).toHaveBeenCalledTimes(5);
    });

    it('runs `nextGen` when clicked', () => {
      pause.simulate('click');
      step.simulate('click');
      expect(props.nextGen).toHaveBeenCalledTimes(6);
    });
  });
});
