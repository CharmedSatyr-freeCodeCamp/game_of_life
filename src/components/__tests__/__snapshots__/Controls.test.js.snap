// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Controls /> component renders correctly 1`] = `
<div
  className="controls"
>
  <button
    className="step disabled"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-step-forward fa-w-14 "
      data-icon="step-forward"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M384 44v424c0 6.6-5.4 12-12 12h-48c-6.6 0-12-5.4-12-12V291.6l-195.5 181C95.9 489.7 64 475.4 64 448V64c0-27.4 31.9-41.7 52.5-24.6L312 219.3V44c0-6.6 5.4-12 12-12h48c6.6 0 12 5.4 12 12z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
     Step
  </button>
  <button
    className="play selected disabled"
    onClick={null}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-play fa-w-14 "
      data-icon="play"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
     Play
  </button>
  <button
    className="pause  "
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-pause fa-w-14 "
      data-icon="pause"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M144 479H48c-26.5 0-48-21.5-48-48V79c0-26.5 21.5-48 48-48h96c26.5 0 48 21.5 48 48v352c0 26.5-21.5 48-48 48zm304-48V79c0-26.5-21.5-48-48-48h-96c-26.5 0-48 21.5-48 48v352c0 26.5 21.5 48 48 48h96c26.5 0 48-21.5 48-48z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
     Pause
  </button>
  <button
    className="clear "
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-times fa-w-11 "
      data-icon="times"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 352 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
     Clear
  </button>
  <button
    className="random "
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-random fa-w-16 "
      data-icon="random"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M504.971 359.029c9.373 9.373 9.373 24.569 0 33.941l-80 79.984c-15.01 15.01-40.971 4.49-40.971-16.971V416h-58.785a12.004 12.004 0 0 1-8.773-3.812l-70.556-75.596 53.333-57.143L352 336h32v-39.981c0-21.438 25.943-31.998 40.971-16.971l80 79.981zM12 176h84l52.781 56.551 53.333-57.143-70.556-75.596A11.999 11.999 0 0 0 122.785 96H12c-6.627 0-12 5.373-12 12v56c0 6.627 5.373 12 12 12zm372 0v39.984c0 21.46 25.961 31.98 40.971 16.971l80-79.984c9.373-9.373 9.373-24.569 0-33.941l-80-79.981C409.943 24.021 384 34.582 384 56.019V96h-58.785a12.004 12.004 0 0 0-8.773 3.812L96 336H12c-6.627 0-12 5.373-12 12v56c0 6.627 5.373 12 12 12h110.785c3.326 0 6.503-1.381 8.773-3.812L352 176h32z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
     Random
  </button>
</div>
`;
